specs/ThrottlingControlScript.md

```diff
@@ -0,0 +1,325 @@
+
+    // User is not interactive, keep webview visible but throttle timers to
+    // 500ms.
+    CHECK_FAILURE(settings2->put_ThrottlingIntervalPreferenceForeground(500));
+}
+
+void ScenarioThrottlingControl::OnUserInteraction()
+{
+    wil::com_ptr<ICoreWebView2Settings> settings;
+    m_webview->get_Settings(&settings);
+    auto settings2 = settings.try_query<ICoreWebView2Settings9>();
+    CHECK_FEATURE_RETURN_EMPTY(settings2);
+
+    // User is interactive again, unthrottle foreground timers.
+    CHECK_FAILURE(settings2->put_ThrottlingIntervalPreferenceForeground(0));
+}
+```
+
+## Unthrottle timers in hidden WebView
+
+Throttling Control APIs allow you to set a custom throttling interval for timers
+on hidden WebViews. For example, if there's logic in your app that runs in
+JavaScript but doesn't need to render content, you can keep the WebView hidden
+and unthrottle its timers.
+
+```C#
+void SetupHiddenWebViewCore()
+{
+    // This WebView2 will remain hidden but needs to keep running timers.
+    // Unthrottle background timers.
+    webView.CoreWebView2.Settings.ThrottlingIntervalPreferenceBackground = 0;
+    // Effectively disable intensive throttling by overriding its timer interval.
+    webView.CoreWebView2.Settings.ThrottlingIntervalPreferenceIntensive = 0;
+    webView.Visibility = System.Windows.Visibility.Hidden;
+}
+
+void DisableHiddenWebViewCore()
+{
+    webView.Visibility = System.Windows.Visibility.Visible;
+    webView.CoreWebView2.Settings.ThrottlingIntervalPreferenceBackground = 1000;
+    webView.CoreWebView2.Settings.ThrottlingIntervalPreferenceIntensive = 60000;
+}
+```
+
+```cpp
+void ScenarioThrottlingControl::SetupHiddenWebViewCore()
+{
+    wil::com_ptr<ICoreWebView2Settings> settings;
+    m_webview->get_Settings(&settings);
+    auto settings2 = settings.try_query<ICoreWebView2Settings9>();
+    CHECK_FEATURE_RETURN_EMPTY(settings2);
+
+    // This WebView2 will remain hidden but needs to keep running timers.
+    // Unthrottle background timers.
+    CHECK_FAILURE(settings2->put_ThrottlingIntervalPreferenceBackground(0));
+    // Effectively disable intensive throttling by overriding its timer interval.
+    CHECK_FAILURE(settings2->put_ThrottlingIntervalPreferenceIntensive(0));
+
+    CHECK_FAILURE(m_appWindow->GetWebViewController()->put_IsVisible(FALSE));
+}
+
+void ScenarioThrottlingControl::DisableHiddenWebViewCore()
+{
+    CHECK_FAILURE(m_appWindow->GetWebViewController()->put_IsVisible(TRUE));
+
+    wil::com_ptr<ICoreWebView2Settings> settings;
+    m_webview->get_Settings(&settings);
+    auto settings2 = settings.try_query<ICoreWebView2Settings9>();
+    CHECK_FEATURE_RETURN_EMPTY(settings2);
+
+    CHECK_FAILURE(settings2->put_ThrottlingIntervalPreferenceBackground(1000));
+    CHECK_FAILURE(settings2->put_ThrottlingIntervalPreferenceIntensive(60000));
+}
+```
+
+## Throttle timers in hosted iframes
+
+Throttling Control APIs allow you to throttle timers in specific frames within
```

Comment:
So every frame is either a 'regular' frame or an 'isolated' frame. Regular frames are controlled by `ThrottlingIntervalPreferenceForeground`/`ThrottlingIntervalPreferenceBackground`/`ThrottlingIntervalPreferenceIntensive`. Isolated frames are controlled by `ThrottlingIntervalPreferenceIsolated`.

Is there a reason why an isolated frame does not have Foreground/Background/Intensive properties? Or is that not an interesting scenario?

Reply:
Valid question - Mostly not an interesting enough scenario to add the complexity. The primary use-case here is an app embedding 3rd party content and wanting to be able to independently limit the performance impact of it. Generally that's something like "low battery, throttle more" or "giving the frame N seconds to run some logic, throttle less".

The case where they'd want to put it in an isolated group while still managing foreground/background/intensive independently of the non-isolated timers would be a niche of a niche that even our most micromanage-y apps wouldn't want that granularity.